types/node/globals.d.ts

```diff
@@ -103,6 +103,84 @@ declare global {
+     *
+     * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage)
+     */
+    interface Storage {
+        /**
+         * Returns the number of key/value pairs.
+         *
+         * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage/length)
+         */
+        readonly length: number;
+        /**
+         * Removes all key/value pairs, if there are any.
+         *
+         * Dispatches a storage event on Window objects holding an equivalent Storage object.
+         *
+         * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage/clear)
+         */
+        clear(): void;
+        /**
+         * Returns the current value associated with the given key, or null if the given key does not exist.
+         *
+         * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage/getItem)
+         */
+        getItem(key: string): string | null;
+        /**
+         * Returns the name of the nth key, or null if n is greater than or equal to the number of key/value pairs.
+         *
+         * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage/key)
+         */
+        key(index: number): string | null;
+        /**
+         * Removes the key/value pair with the given key, if a key/value pair with the given key exists.
+         *
+         * Dispatches a storage event on Window objects holding an equivalent Storage object.
+         *
+         * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage/removeItem)
+         */
+        removeItem(key: string): void;
+        /**
+         * Sets the value of the pair identified by key to value, creating a new key/value pair if none existed for key previously.
+         *
+         * Throws a "QuotaExceededError" DOMException exception if the new value couldn't be set. (Setting could fail if, e.g., the user has disabled storage for the site, or if the quota has been exceeded.)
+         *
+         * Dispatches a storage event on Window objects holding an equivalent Storage object.
+         *
+         * [MDN Reference](https://developer.mozilla.org/docs/Web/API/Storage/setItem)
+         */
+        setItem(key: string, value: string): void;
+    }
+
+    var Storage: typeof globalThis extends { onmessage: any; Storage: infer T } ? T
+        : {
+            prototype: Storage;
+            new(): Storage;
+        };
+
+    /**
+     * A browser-compatible implementation of [`localStorage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage).
+     * Data is stored unencrypted in the file specified by the `--localstorage-file` CLI flag.
+     * Any modification of this data outside of the Web Storage API is not supported.
+     * Enable this API with the `--experimental-webstorage` CLI flag.
+     * @since v22.4.0
+     */
+    var localStorage: Storage;
```

Comment:
@Semigradsky what are you thoughts about this suggested change?

Since these globals are only available at runtime when the `--experimental-webstorage` is passed, I'd say it's confusing to provide them without needing runtime checks.

```suggestion
    var localStorage: Storage | undefined;
```

Reply:
The convention within @types/node (_mostly_) is not to type features that live behind flags as possibly `undefined`.

Not only does this force type checking/assertions on projects that know their own build environment, but for global variables, it's just incorrect: it's not the case that the variable is potentially undefined, it's that it potentially doesn't exist, and those are not equivalent.

```ts
declare var localStorage: Storage | undefined;

// TypeScript tells me I should check that this isn't undefined...
if (localStorage) {
  localStorage.clear();
}
```

The TypeScript compiler is perfectly happy here, which makes it a rather jarring experience when you run this and encounter `Uncaught ReferenceError: localStorage is not defined`.

Reply:
Ophf 🤔 you're right, that case is not covered. I wondered if using `never` would be better, but it doesn't seem to be the case.

At the very least, this could have an [experimental](https://tsdoc.org/pages/tags/experimental/) tag? Not that I'd expect great IDE support for highlighting that 😔

Reply:
One thing to note is that you still need to end up in a position where loading the Node types and the DOM types together doesn't fall over; DT should be testing that already, and I would suspect that adding `| undefined` would probably cause errors.

Reply:
Another alternative to keep things sane around types for experimental features would be to separate these experimental types into a separate file, such that users can opt into the feature via a simple top-level import:

```typescript
import type "node/experimental-webstorage";
```

Is that something which has been discussed already?

Reply:
I don't think it's a good idea to introduce imports which would crash at runtime; there's no way to enforce that an import is only used as a type import.

IMO this seems like a lot of work for not too much benefit; I'll note that the DOM types have some cases like this already for features that are only enabled in secure contexts, and those aren't even experimental.

Reply:
> I don't think it's a good idea to introduce imports which would crash at runtime;

I agree, but that's the point of my initial comment. In its current form the `localStorage` is declared but not even defined, resulting in the `Uncaught ReferenceError: localStorage is not defined` @Renegade334 mentioned above. Putting this in a separate file users would at least have to opt-into getting these experimental types declared on the global namespace.